### PR TITLE
X509Certificates.Tests: make tests pass when rsa+sha1 signing is not supported.

### PR DESCRIPTION
**Pull request by tmds**
@vcsjones @bartonjs ptal.

This should be the last of the test suites that need a change.

### CODE REVIEW

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs [hunk 1]
```diff
@@ -169,7 +169,7 @@ public static void TestDispose()
             Assert.Equal(IntPtr.Zero, chain.ChainContext);
         }
 
-        [Fact]
+        [ConditionalFact(typeof(SignatureSupport), nameof(SignatureSupport.SupportsRsaSha1Signatures))]
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:172]: This test is not using RSA+SHA1, it's using DSA(+SHA1).

File: src/libraries/System.Security.Cryptography.X509Certificates/tests/RevocationTests/DynamicRevocationTests.cs [hunk 2]
```diff
@@ -364,7 +364,7 @@ public static void RevokeEverything(PkiOptions pkiOptions)
                 });
         }
 
-        [Theory]
+        [ConditionalTheory(typeof(SignatureSupport), nameof(SignatureSupport.SupportsRsaSha1Signatures))]
```
Comment (vcsjones) [inline src/libraries/System.Security.Cryptography.X509Certificates/tests/RevocationTests/DynamicRevocationTests.cs:367]: This test fails for me with:

<pre>
System.AggregateException : One or more errors occurred. (Assert.Equal() Failure
                             ↓ (pos 2)
Expected: [NoError, NoError, RevocationStatusUnknown | OfflineRevocati···
Actual:   [NoError, NoError, Revoked]
</pre>

Are we sure we are disabling the test on that platform for the right reason?

Comment (tmds) [reply src/libraries/System.Security.Cryptography.X509Certificates/tests/RevocationTests/DynamicRevocationTests.cs:367]: I'll go back and look at the exceptions. Also for https://github.com/dotnet/runtime/pull/70821#discussion_r899176346.

File: src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs [hunk 3]
```diff
@@ -558,7 +563,7 @@ public static void BuildChain_WithApplicationPolicy_NoMatch()
             }
         }
 
-        [Fact]
+        [ConditionalFact(typeof(SignatureSupport), nameof(SignatureSupport.SupportsRsaSha1Signatures))]
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:566]: Does this need to be conditionally run? Looks like it could just use that boolean as guards around the expected chain success?

File: src/libraries/System.Security.Cryptography.X509Certificates/tests/RevocationTests/DynamicRevocationTests.cs (hunk 2 above)
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.X509Certificates/tests/RevocationTests/DynamicRevocationTests.cs:367]: This test doesn't appear to have anything to do with SHA-1, it should not be disabled

File: src/libraries/System.Security.Cryptography.X509Certificates/tests/RevocationTests/DynamicRevocationTests.cs [hunk 4]
```diff
@@ -449,7 +449,7 @@ public static void RevokeEndEntity_IssuerUnrelatedOcsp(PkiOptions pkiOptions)
...
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.X509Certificates/tests/RevocationTests/DynamicRevocationTests.cs:452]: Similarly, I don't see where RSA+SHA1 is coming into play here

File: src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs (hunk 1 above)
Comment (tmds) [reply src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:172]: The chain fails to build.
```
<test name="System.Security.Cryptography.X509Certificates.Tests.ChainTests.TestResetMethod" type="System.Security.Cryptography.X509Certificates.Tests.ChainTests" method="TestResetMethod" time="0.0028312" result="Fail">
<failure exception-type="Xunit.Sdk.TrueException">
<message>
Chain built validly\nExpected: True\nActual: False
</message>
<stack-trace>
at System.Security.Cryptography.X509Certificates.Tests.ChainTests.TestResetMethod() in /home/tester/runtime/src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:line 194 at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor) at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr) in /home/tester/runtime/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodInvoker.cs:line 69
</stack-trace>
</failure>
</test>
```

File: src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs (hunk 3 above)
Comment (tmds) [reply src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:566]: Yes, I'll do that instead.

**Comment by tmds** [comment]
@bartonjs @vcsjones some test failures are indeed not about the SHA1, the flags differ from what the tests expect:

- For `RevokeEndEntity_IssuerUnrelatedOcsp`:

In most cases, the assert fails like:
```
E: [NoError, NoError, RevocationStatusUnknown | OfflineRevocati···
A: [NoError, NoError, Revoked]
```
In some cases `RevocationStatusUnknown | OfflineRevocati` shows up instead of `NoError` in the second item:
```
E: [NoError, NoError, RevocationStatusUnknown | Offlin···
A: [NoError, RevocationStatusUnknown | OfflineRevocati···
```
- For `RevokeEndEntity_RootUnrelatedOcsp`:
```
E: [NoError, RevocationStatusUnknown | OfflineRevocati···
A: [NoError, Revoked, RevocationStatusUnknown | Offlin···
```

**Comment by vcsjones** [comment]
@tmds those look like what I saw in https://github.com/dotnet/runtime/issues/68911. I haven't had a chance to dig in to those or try to reproduce it outside .NET.

File: src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs (hunk 1 above)
Comment (bartonjs) [reply src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:172]: OK... but.... why?

If the same setting in RHEL is disabling DSA+SHA1, then we have a reason, but then we should make a second SignatureSupport for SupportsDsaSha1Signatures.  And then we shouldn't turn off the test, we should expect false and a particular set of flags.

It's quite possible this is the only test that's building chains with DSA, if it's the only place failing.

Comment (tmds) [reply src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:172]: Looked a bit closer:
```
DSA.Create().SignData(Array.Empty<byte>(), HashAlgorithmName.SHA1)
```
doesn't throw `CryptographicException`. So DSA+SHA1 signatures seem to be supported.

For this test, `chain.ChainStatus[0].Status` is `NotSignatureValid`.

**Comment by danmoseley** [comment]
@bartonjs can you help advance this stale PR?

**Comment by bartonjs** [comment]
Any tips on getting into a state where I can locally run it?

After spending a lot of time getting a fedora 36 build working I learned that docker's fedora:latest is older than this policy, and I needed fedora 37.  On Fedora 37 I seem to have the choice of having clang crash while building dotnet/main, or having the build fail while building dotnet/release/6.0.

So, some way to get to a place where I can build and run would be great, or someone with a "uh, you're doing it wrong, just ..." on sending some runnable builds into a fed37 docker container so I can debug.

**Comment by vcsjones** [comment]
@bartonjs I don't know how set you are on using Fedora but here is a Dockerfile for CentOS 9 that gets all of the dependencies installed to running the X509 tests.

```Dockerfile
FROM quay.io/centos/centos:stream9

# Dependencies! Yay!
RUN dnf -y groupinstall 'Development Tools' && \
        dnf install -y dnf-plugins-core epel-release && \
        dnf repolist --all && \
        dnf config-manager --set-enabled crb && \
        dnf install -y \
            clang \
            cmake \
            git \
            glibc-langpack-en \
            hostname \
            krb5-devel \
            libicu-devel \
            lld \
            llvm \
            make \
            openssl-devel \
            python3 \
            tar \
            util-linux \
            zlib-devel \
            lttng-ust-devel \
            libunwind-devel

RUN mkdir /projects && \
    cd /projects && \
    git clone https://github.com/dotnet/runtime runtime

WORKDIR /projects/runtime

# Need to do the initial build allowing SHA1 signatures.
RUN OPENSSL_ENABLE_SHA1_SIGNATURES=1 ./build.sh -rc release -s clr+libs

# Build the tests allowing SHA1 signatures, too.
RUN OPENSSL_ENABLE_SHA1_SIGNATURES=1 ./dotnet.sh build src/libraries/System.Security.Cryptography.X509Certificates/tests

# Actually run the tests _without_ allowing SHA1.
RUN ./dotnet.sh test --no-restore src/libraries/System.Security.Cryptography.X509Certificates/tests
```

If you're still stuck, you know where to find me. We can find some time to pair on it.

**Comment by bartonjs** [comment]
Hm, same clang crash as I saw on Fedora 36 and 37:

```
  PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
  Stack dump:
  0.    Program arguments: /usr/bin/clang++-14 -DDEBUGGING_SUPPORTED -DDISABLE_CONTRACTS -DEnC_SUPPORTED -DFEATURE_ARRAYSTUB_AS_IL -DFEATURE_BASICFREEZE -DFEATURE_CODE_VERSIONING -DFEATURE_COLLECTIBLE_TYPES -DFEATURE_COMWRAPPERS -DFEATURE_CORECLR -DFEATURE_DBGIPC_TRANSPORT_DI -DFEATURE_DBGIPC_TRANSPORT_VM -DFEATURE_DEFAULT_INTERFACES -DFEATURE_EH_FUNCLETS -DFEATURE_EVENTSOURCE_XPLAT -DFEATURE_EVENT_TRACE -DFEATURE_HIJACK -DFEATURE_HW_INTRINSICS -DFEATURE_ICASTABLE -DFEATURE_INSTANTIATINGSTUB_AS_IL -DFEATURE_ISYM_READER -DFEATURE_MANAGED_ETW -DFEATURE_MANAGED_ETW_CHANNELS -DFEATURE_MANUALLY_MANAGED_CARD_BUNDLES -DFEATURE_MULTICASTSTUB_AS_IL -DFEATURE_MULTICOREJIT -DFEATURE_MULTIREG_RETURN -DFEATURE_NO_HOST -DFEATURE_ON_STACK_REPLACEMENT -DFEATURE_PAL_ANSI -DFEATURE_PERFMAP -DFEATURE_PERFTRACING -DFEATURE_PGO -DFEATURE_PORTABLE_SHUFFLE_THUNKS -DFEATURE_PROFAPI_ATTACH_DETACH -DFEATURE_READYTORUN -DFEATURE_REJIT -DFEATURE_REMOTE_PROC_MEM -DFEATURE_SIMD -DFEATURE_STANDALONE_GC -DFEATURE_STUBS_AS_IL -DFEATURE_SVR_GC -DFEATURE_SYMDIFF -DFEATURE_TIERED_COMPILATION -DFEATURE_USE_ASM_GC_WRITE_BARRIERS -DFEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP -DHOST_64BIT -DHOST_AMD64 -DHOST_UNIX -DJIT_BUILD -DNDEBUG -DPROFILING_SUPPORTED -DSELF_NO_HOST -DTARGET_64BIT -DTARGET_AMD64 -DTARGET_UNIX -DUNICODE -DUNIX_AMD64_ABI -DUNIX_AMD64_ABI_ITF -DURTBLDENV_FRIENDLY=Retail -D_BLD_CLR -D_FILE_OFFSET_BITS=64 -D_SECURE_SCL=0 -D_UNICODE -Dclrjit_EXPORTS -I/projects/runtime/artifacts/obj/coreclr/Linux.x64.Release/jit -I/projects/runtime/src/coreclr/jit -I/projects/runtime/src/native -I/projects/runtime/src/coreclr/pal/prebuilt/inc -I/projects/runtime/artifacts/obj -I/projects/runtime/src/coreclr/pal/inc -I/projects/runtime/src/coreclr/pal/inc/rt -I/projects/runtime/src/coreclr/pal/src/safecrt -I/projects/runtime/src/coreclr/inc -I/projects/runtime/src/coreclr/debug/inc -I/projects/runtime/src/coreclr/debug/inc/amd64 -I/projects/runtime/src/coreclr/debug/inc/dump -I/projects/runtime/src/coreclr/md/inc -I/projects/runtime/src/coreclr/classlibnative/bcltype -I/projects/runtime/src/coreclr/classlibnative/cryptography -I/projects/runtime/src/coreclr/classlibnative/inc -I/projects/runtime/artifacts/obj/coreclr/Linux.x64.Release/inc -I/projects/runtime/src/coreclr/hosts/inc -I/projects/runtime/src/coreclr/minipal -I/projects/runtime/src/coreclr/pal/inc/rt/cpp -I/projects/runtime/src/coreclr/nativeresources -I/projects/runtime/src/coreclr/jit/./jitstd -I/projects/runtime/src/coreclr/jit/../inc -O3 -DNDEBUG -fPIC -fPIC -O3 -g -Wall -Wno-null-conversion -fno-omit-frame-pointer -fms-extensions -fwrapv -fstack-protector-strong -Werror -Wno-unused-variable -Wno-unused-value -Wno-unused-function -Wno-tautological-compare -Wno-unknown-pragmas -Wimplicit-fallthrough -Wno-invalid-offsetof -Wno-unused-but-set-variable -ffp-contract=off -Wno-unknown-warning-option -ferror-limit=4096 -Wno-unused-private-field -Wno-constant-logical-operand -Wno-pragma-pack -Wno-incompatible-ms-struct -Wno-reserved-identifier -fsigned-char -fvisibility=hidden -nostdinc -flto -fprofile-instr-use=/root/.nuget/packages/optimization.linux-x64.pgo.coreclr/1.0.0-prerelease.22375.7/data/coreclr.profdata -Wno-profile-instr-out-of-date -Wno-profile-instr-unprofiled -std=gnu++11 -Winvalid-pch -Xclang -include-pch -Xclang /projects/runtime/artifacts/obj/coreclr/Linux.x64.Release/jit/CMakeFiles/clrjit.dir/cmake_pch.hxx.pch -Xclang -include -Xclang /projects/runtime/artifacts/obj/coreclr/Linux.x64.Release/jit/CMakeFiles/clrjit.dir/cmake_pch.hxx -MD -MT jit/CMakeFiles/clrjit.dir/gentree.cpp.o -MF CMakeFiles/clrjit.dir/gentree.cpp.o.d -o CMakeFiles/clrjit.dir/gentree.cpp.o -c /projects/runtime/src/coreclr/jit/gentree.cpp
  1.    /projects/runtime/src/coreclr/jit/gentree.cpp:4300:1: current parser token 'unsigned'
  2.    /projects/runtime/src/coreclr/jit/gentree.cpp:3905:16: LLVM IR generation of declaration 'Compiler::gtMarkAddrMode'
  3.    /projects/runtime/src/coreclr/jit/gentree.cpp:3905:16: Generating code for declaration 'Compiler::gtMarkAddrMode'
   #0 0x00007fcae5397715 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/lib64/libLLVM-14.so+0xb0f715)
   #1 0x00007fcae53956d4 llvm::sys::RunSignalHandlers() (/lib64/libLLVM-14.so+0xb0d6d4)
   #2 0x00007fcae5396d0b llvm::sys::CleanupOnSignal(unsigned long) (/lib64/libLLVM-14.so+0xb0ed0b)
   #3 0x00007fcae52c9909 (/lib64/libLLVM-14.so+0xa41909)
   #4 0x00007fcae44a5320 __restore_rt (/lib64/libc.so.6+0x55320)
   #5 0x00007fcae44f1adc __pthread_kill_implementation (/lib64/libc.so.6+0xa1adc)
   #6 0x00007fcae44a5276 gsignal (/lib64/libc.so.6+0x55276)
   #7 0x00007fcae44787f3 abort (/lib64/libc.so.6+0x287f3)
   #8 0x00007fcaea45c5c8 (/lib64/libclang-cpp.so.14+0x8ac5c8)
   #9 0x00007fcaeb85b867 (/lib64/libclang-cpp.so.14+0x1cab867)
  #10 0x00007fcaeb861f76 (/lib64/libclang-cpp.so.14+0x1cb1f76)
  #11 0x00007fcaeb8614de (/lib64/libclang-cpp.so.14+0x1cb14de)
  #12 0x00007fcaeb85dfa6 (/lib64/libclang-cpp.so.14+0x1cadfa6)
  #13 0x00007fcaeb8614de (/lib64/libclang-cpp.so.14+0x1cb14de)
  #14 0x00007fcaeb86f9f0 clang::CodeGen::CodeGenPGO::computeRegionCounts(clang::Decl const*) (/lib64/libclang-cpp.so.14+0x1cbf9f0)
  #15 0x00007fcaeb86fdeb clang::CodeGen::CodeGenPGO::assignRegionCounters(clang::GlobalDecl, llvm::Function*) (/lib64/libclang-cpp.so.14+0x1cbfdeb)
  #16 0x00007fcaeb81b2ad clang::CodeGen::CodeGenFunction::GenerateCode(clang::GlobalDecl, llvm::Function*, clang::CodeGen::CGFunctionInfo const&) (/lib64/libclang-cpp.so.14+0x1c6b2ad)
  #17 0x00007fcaeb84fde4 clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/lib64/libclang-cpp.so.14+0x1c9fde4)
  #18 0x00007fcaeb84c05b clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/lib64/libclang-cpp.so.14+0x1c9c05b)
  #19 0x00007fcaeb8534f7 (/lib64/libclang-cpp.so.14+0x1ca34f7)
  #20 0x00007fcaeb8baf21 (/lib64/libclang-cpp.so.14+0x1d0af21)
  #21 0x00007fcaeb7fbd17 (/lib64/libclang-cpp.so.14+0x1c4bd17)
  #22 0x00007fcaea5b0144 clang::ParseAST(clang::Sema&, bool, bool) (/lib64/libclang-cpp.so.14+0xa00144)
  #23 0x00007fcaebefb709 clang::FrontendAction::Execute() (/lib64/libclang-cpp.so.14+0x234b709)
  #24 0x00007fcaebe89ad9 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/lib64/libclang-cpp.so.14+0x22d9ad9)
  #25 0x00007fcaebf67dc2 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/lib64/libclang-cpp.so.14+0x23b7dc2)
  #26 0x000055ac5d0a563c cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/usr/bin/clang++-14+0x1663c)
  #27 0x000055ac5d0a225f (/usr/bin/clang++-14+0x1325f)
  #28 0x00007fcaebb91209 (/lib64/libclang-cpp.so.14+0x1fe1209)
  #29 0x00007fcae52c9645 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/lib64/libLLVM-14.so+0xa41645)
  #30 0x00007fcaebb91ab9 (/lib64/libclang-cpp.so.14+0x1fe1ab9)
  #31 0x00007fcaebb64ddb clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&) const (/lib64/libclang-cpp.so.14+0x1fb4ddb)
  #32 0x00007fcaebb6558e clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) const (/lib64/libclang-cpp.so.14+0x1fb558e)
  #33 0x00007fcaebb70192 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) (/lib64/libclang-cpp.so.14+0x1fc0192)
  #34 0x000055ac5d09fe56 main (/usr/bin/clang++-14+0x10e56)
  #35 0x00007fcae4490440 __libc_start_call_main (/lib64/libc.so.6+0x40440)
  #36 0x00007fcae44904f0 __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x404f0)
  #37 0x000055ac5d0a1695 _start (/usr/bin/clang++-14+0x12695)
  clang-14: error: clang frontend command failed with exit code 134 (use -v to see invocation)
  clang version 14.0.5 (Red Hat 14.0.5-1.el9)
  Target: x86_64-redhat-linux-gnu
  Thread model: posix
  InstalledDir: /usr/bin
  clang-14: note: diagnostic msg:
  ********************

  PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
  Preprocessed source(s) and associated run script(s) are located at:
  clang-14: note: diagnostic msg: /tmp/gentree-78d359.cpp
  clang-14: note: diagnostic msg: /tmp/gentree-78d359.sh
  clang-14: note: diagnostic msg:

  ********************
  make[2]: *** [jit/CMakeFiles/clrjit.dir/build.make:573: jit/CMakeFiles/clrjit.dir/gentree.cpp.o] Error 134
  make[1]: *** [CMakeFiles/Makefile2:3346: jit/CMakeFiles/clrjit.dir/all] Error 2
  make[1]: *** Waiting for unfinished jobs....
  /usr/include/c++/11/bits/stl_vector.h:1045: std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](std::vector<_Tp, _Alloc>::size_type) [with _Tp = long unsigned int; _Alloc = std::allocator<long unsigned int>; std::vector<_Tp, _Alloc>::reference = long unsigned int&; std::vector<_Tp, _Alloc>::size_type = long unsigned int]: Assertion '__n < this->size()' failed.
```

**Comment by vcsjones** [comment]
@bartonjs 

Huh. I don't hit any compiler asserts. I nuked my docker image cache and pulled everything from latest. Maybe it's because I am on ARM64.

Looks like @omajid reported this: https://github.com/llvm/llvm-project/issues/52991

Following more links, we end up back here: https://github.com/dotnet/runtime/issues/65676#issuecomment-1047194765

So it looks like you can avoid your compiler crashes if you build with `/p:NoPgoOptimize=true`.

**Comment by bartonjs** [comment]
Thanks @vcsjones.

Achievement Unlocked: Use docker to use an OS different than the one that booted.
Achievement Unlocked: Use docker with a Dockerfile.

**Comment by bartonjs** [comment]
Ok, by running xunit with `-verbose` (and `-method  System.Security.Cryptography.X509Certificates.Tests.ChainTests.TestResetMethod`) so I can tell when the test starts, breaking on `CryptoNative_X509VerifyCert` (which has a few hits before the test starts, so continue past them), and then enabling a breakpoint for `ERR_set_error`, I see that TestResetMethod fails at

```
  * frame #0: 0x00007fbeca317ea0 libcrypto.so.3`ERR_set_error
    frame #1: 0x00007fbeca343811 libcrypto.so.3`do_sigver_init + 2433
    frame #2: 0x00007fbeca343b6c libcrypto.so.3`EVP_DigestVerifyInit + 28
    frame #3: 0x00007fbeca267e6b libcrypto.so.3`ASN1_item_verify_ctx + 571
    frame #4: 0x00007fbeca2681fa libcrypto.so.3`ASN1_item_verify_ex + 74
    frame #5: 0x00007fbeca3dacb2 libcrypto.so.3`X509_verify + 114
    frame #6: 0x00007fbeca3dad8c libcrypto.so.3`internal_verify + 188
    frame #7: 0x00007fbeca3dc6be libcrypto.so.3`verify_chain + 3054
    frame #8: 0x00007fbeca3dcc42 libcrypto.so.3`X509_verify_cert + 194
    frame #9: 0x00007fbeca6a67f5 libSystem.Security.Cryptography.Native.OpenSsl.so`CryptoNative_X509VerifyCert(ctx=0x00007fbebc009970) at pal_x509.c:286:12
```

Looking at [the patch](https://gitlab.com/redhat/centos-stream/rpms/openssl/-/commit/78fb78d30755ae18fdaef28ef392f4e67c662ff6), the build blocks any asymmetric signature algorithm from using SHA-1... if it goes through EVP_DigestVerifyInit.  So here it's blocking DSA+SHA-1.

Our normal DSA tests aren't failing, though, because they don't go through `EVP_DigestVerify*`.  Even if they went through `EVP_*` routines at all, they wouldn't use EVP_DigestVerify, because we base on VerifyHash, with the hashes computed previously, and since DSA doesn't need to know the hash algorithm during the sign or verify step, we probably wouldn't need to call `EVP_PKEY_ctx_set_md`.

So, basically, the current patch to OpenSSL that has heretofore been described as blocking RSA+SHA-1 blocks RSA+SHA-1 ... and also blocks DSA+SHA-1 or ECDSA+SHA-1 when the digest isn't done ahead of time.  Since the chain engine internally uses `EVP_DigestVerify*` that means it blocks certificate chains making use of ECDSA+SHA-1 and DSA+SHA-1.

If that's intentional, then I guess the only change off of the PR is to rename the condition from `SupportsRsaSha1Signatures` to `SupportsX509Sha1Signatures` (RSA => X509 in the condition).  If that's not something that the RHEL/CentOS/Fedora patch was expected to do, then I guess we keep holding this up until the non-RSA tests are passing.

**Comment by omajid** [comment]
From [RHEL 9 docs](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html/considerations_in_adopting_rhel_9/assembly_security_considerations-in-adopting-rhel-9

> In RHEL 9, SHA-1 usage for signatures is restricted in the DEFAULT system-wide cryptographic policy. Except for HMAC, SHA-1 is no longer allowed in TLS, DTLS, SSH, IKEv2, DNSSEC, and Kerberos protocols. Individual applications not controlled by the RHEL system-wide crypto policies are also moving away from using SHA-1 hashes in RHEL 9. 

It sounds like it's expected that SHA-1 isn't going to work with DSA or ECDSA either?

**Comment by bartonjs** [comment]
> It sounds like it's expected that SHA-1 isn't going to work with DSA or ECDSA either?

Well, that blurb does make it sound like SHA-1 won't work with DSA or ECDSA in "TLS, DTLS, SSH, IKEv2, DNSSEC, and Kerberos protocols".  The implementation as it currently stands is inconsistent.

* RSA+SHA-1 is always blocked.
* DSA+SHA-1 is blocked if you ask the signature generator or verifier to do the hashing for you.
* DSA+SHA-1 is not blocked if you precompute the hash for the generator or verifier.
* ECSDA behaves just like DSA.

So on RHEL9 the DSA+SHA-1 direct usage tests and ECDSA+SHA-1 direct usage tests (assuming we have some) in .NET are passing, but if they're used in an X.509 chain they don't work (unless they were the self-signed signature of a trusted root). That makes it a little hard to know how to proceed.

* If "it is what it is", then we (well, probably Tom) rename(s) the condition in this test library from RsaSha1 to X509Sha1, and we're done.
* If Red Hat wants to change to always block DSA+SHA-1 and ECDSA+SHA-1 we have more tests to disable and should change all of our conditions to be called "SupportsSha1Signature" or something.
* If Red Hat wants to change to only block RSA+SHA-1 then we get to remove some of the suppressions in this PR.
* If Red Hat wants to change to say "well, 99.999% of DSA usage is SHA-1, and people who need to verify it need to verify it, so we'll block RSA+SHA-1 and ECDSA+SHA-1, but DSA+SHA-1 gets to slide" then, well, we have different changes to make.

**Comment by omajid** [comment]
Hey @neverpanic @simo5, can you share any details on how you expect SHA1 + RSA/DSA/ECDSA to behave on RHEL 9 given the questions above?

**Comment by simo5** [comment]
@omajid in default configuration any signature using SHA-1 is denied.

Given that, if you pre-compute the hash, we do not know what the content is, we can't block those API only for SHA-1, so we do not block them.
We recommend application to *not8 use those APIs as they are generally blocked entirely (f0r any hash function) in FIPS mode.

I would suggest skipping SHA-1 signature related test on RHEL-9, and in general plan a deprecation, SHA-1 *is* broken for signatures and upstream should cease using it too.

**Comment by neverpanic** [comment]
You're essentially hitting what's described in the patch header for the [patch you linked earlier](https://gitlab.com/redhat/centos-stream/rpms/openssl/-/commit/78fb78d30755ae18fdaef28ef392f4e67c662ff6). I'm quoting below:

> For signatures created using the EVP_PKEY API, this is a best-effort check that will deny signatures in cases where the digest algorithm is known. This means, for example, that that following steps will still work:
>
> $> openssl dgst -sha1 -binary -out sha1 infile
> $> openssl pkeyutl -inkey key.pem -sign -in sha1 -out sha1sig
> $> openssl pkeyutl -inkey key.pem -verify -sigfile sha1sig -in sha1
>
> whereas these will not:
>
> $> openssl dgst -sha1 -binary -out sha1 infile
> $> openssl pkeyutl -inkey kem.pem -sign -in sha1 -out sha1sig -pkeyopt digest:sha1
> $> openssl pkeyutl -inkey kem.pem -verify -sigfile sha1sig -in sha1 -pkeyopt digest:sha1
>
> This happens because in the first case, OpenSSL's signature implementation does not know that it is signing a SHA1 hash (it could be signing arbitrary data).

We would also have disabled these signatures if there was a way for us to determine that what is being signed is in fact a SHA-1 hash. We can't block the computation of a SHA-1 hash, because there are still legitimate uses for SHA-1 outside of signatures. In the signature API, we block SHA-1 if we recognize it, which is pretty much only the case for RSASSA-PKCS1-v1_5 signatures, because those contain an ASN.1 structure that encodes the used hash algorithms, where most other signature schemes (such as DSA, ECDSA, and even RSASSA-PSS) do not.

File: src/libraries/System.Security.Cryptography.X509Certificates/tests/SignatureSupport.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Xml;
+using Xunit;
+
+namespace System.Security.Cryptography.X509Certificates.Tests
+{
+    public class SignatureSupport
+    {
+        public static bool SupportsRsaSha1Signatures { get; } =
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.X509Certificates/tests/SignatureSupport.cs:11]: Let's rename it at this layer so it justifies suppressing DSA.

```suggestion
        public static bool SupportsX509Sha1Signatures { get; } =
```

**Comment by bartonjs** [comment]
All of the failures reported thus far have had console logs saying all the tests passed, and none have been in the library involved in this test-only change.

I've seen that BuildChain_WithCertificatePolicy_Match still runs on systems we expect it to, so we're not too generous with the skip condition.

Going ahead and merging so that the release branches start off happier on CentOS9.

**Comment by tmds** [comment]
@bartonjs thank you for finishing this PR!
